### PR TITLE
Configurable data documentation sites

### PR DESCRIPTION
**Pull request by eugmandel**
A data documentation site consists of HTML pages for expectation suites, profiling and validation results, and an index.html page that links to all the pages.

It is possible to have multiple sites within one project. 

The behavior of each site is controlled by configuration in great_expectations.yml file.

The CLI "init" command precreates two data documentation sites for a new project: 
1. "local_site" that visualizes all expectation suites, profiling results and validation results in "uncommitted" directory. This removes the need to view JSON files.
2. "team_site" that is a placeholder for a site that a team can use to maintain the "shared truth" of the project.

    Users can specify
    * which datasources to document (by default, all)
    * whether to include expectations, validations and profiling results sections
    * where the expectations and validations should be read from (filesystem or S3)
    * where the HTML files should be written (filesystem or S3)
    * which renderer and view class should be used to render each section

Overview of the feature:
https://www.loom.com/share/a3c510159ad04cb8923c43c3cf032163

### CODE REVIEW

File: great_expectations/cli/cli.py [hunk 1]
```diff
@@ -259,16 +259,24 @@ def profile(datasource_name, data_assets, profile_all_data_assets, directory):
 @cli.command()
 @click.option('--directory', '-d', default="./great_expectations",
               help='The root of a project directory containing a great_expectations/ config.')
-def documentation(directory):
+@click.option('--site_name', '-s',
+              help='The site for which to generate documentation. See data_docs section in great_expectations.yml')
+@click.option('--data_asset_name', '-dan',
```
Comment (abegong) [inline great_expectations/cli/cli.py:264]: ❤️

File: great_expectations/data_context/data_context.py [hunk 2]
```diff
@@ -191,6 +218,17 @@ def plugins_directory(self):
         """The directory in which custom plugin modules should be placed."""
         return self._plugins_directory
 
+    @property
+    def expectations_directory(self):
```
Comment (abegong) [inline great_expectations/data_context/data_context.py:222]: This isn't really the focus of this release, but I'm really curious about how custom plugin modeuls work. Do we have an example anywhere?

File: great_expectations/data_context/data_context.py [hunk 3]
```diff
@@ -316,7 +353,7 @@ def get_validation_doc_filepath(self, data_asset_name, expectation_suite_name):
             path (str): Path to the location
 
         """
-        # TODO: this path should be configurable or parameterized to support expectation suite and profiling results docs
+        # TODO: this path should be configurable or parameterized to support descriptive and prescriptive docs
```
Comment (abegong) [inline great_expectations/data_context/data_context.py:356]: ? I thought we were getting rid of this terminology.

File: great_expectations/data_context/data_context.py [hunk 4]
```diff
@@ -1098,25 +1135,25 @@ def register_validation_results(self, run_id, validation_results, data_asset=Non
             )
 
         expectation_suite_name = validation_results["meta"].get("expectation_suite_name", "default")
-        if "result_store" in self._project_config:
-            result_store = self._project_config["result_store"]
-            if isinstance(result_store, dict) and "filesystem" in result_store:
+        if self.validations_store:
+            validations_store = self.validations_store
+            if isinstance(validations_store, dict) and validations_store["type"] == "filesystem":
```
Comment (abegong) [inline great_expectations/data_context/data_context.py:1140]: This looks like a great candidate for a `LooselyTypedDotDict`

File: great_expectations/data_context/data_context.py [hunk 5]
```diff
@@ -1403,21 +1440,156 @@ def _compile(self):
 
         self._compiled = True
 
-    def get_validation_result(self, data_asset_name, expectation_suite_name="default", run_id=None, failed_only=False):
+    def write_resource(
+            self,
+            resource,  # bytes
```
Comment (abegong) [inline great_expectations/data_context/data_context.py:1445]: What is a `resource`?

File: great_expectations/data_context/data_context.py [hunk 6]
```diff
@@ -1403,21 +1440,156 @@ def _compile(self):
...
+        }
+        """
+        if validations_store is None:
+            validations_store = self.validations_store
+        else:
+            validations_store = self._normalize_store_path(validations_store)
+
+        validation_results = {}
+
+        if validations_store["type"] == "filesystem":
```
Comment (abegong) [inline great_expectations/data_context/data_context.py:1535]: It feels like we're going to want to create a separate `Store` class to handle this kind of IO logic across different types of stores. #later

File: tests/render/test_data_documentation_site_builder.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+from great_expectations.render.renderer.site_builder import SiteBuilder
+
+
+
+def test_cli_profile(titanic_data_context):
+    print("BOOO")
+
+    titanic_data_context.profile_datasource(titanic_data_context.list_datasources()[0]["name"])
+
+    site_config = {
```
Comment (abegong) [inline tests/render/test_data_documentation_site_builder.py:12]: yah, this should def. be a `LooselyTypedDotDict`

File: tests/render/test_data_documentation_site_builder.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import os
+import pytest
+from great_expectations.render.renderer.site_builder import SiteBuilder
+
+
+
+def test_cli_profile(titanic_data_context):
+    print("BOOO")
```
Comment (abegong) [inline tests/render/test_data_documentation_site_builder.py:8]: 😆

**Review by abegong — APPROVED** [review]
LGTM.

I added comments, but nothing blocking.

I see what you mean about tests being a little bit thin. I bet adding the next 2 or 3 of the most urgent unit tests (e.g. internal methods for Stores and site building) will pay off strongly in next month.

File: great_expectations/data_context/data_context.py (hunk 3 above)
Comment (eugmandel) [reply great_expectations/data_context/data_context.py:356]: @abegong You are right - we will not use it in the docs and should scrub it in the comments too.

File: great_expectations/data_context/data_context.py (hunk 4 above)
Comment (eugmandel) [reply great_expectations/data_context/data_context.py:1140]: @abegong It does - if it does not make it into 0.7.6, I'll create a task for 0.7.7.

File: great_expectations/data_context/data_context.py (hunk 5 above)
Comment (eugmandel) [reply great_expectations/data_context/data_context.py:1445]: @abegong technically a resource is a string of bytes. Usually it is an HTML document. This API allows to plug in writing to a local filesystem, S3 and whatever medium we will find useful in the future.

File: great_expectations/data_context/data_context.py (hunk 6 above)
Comment (eugmandel) [reply great_expectations/data_context/data_context.py:1535]: @abegong yes, this is what @jcampbell  and I discussed, but the first pass of the implementation is a just a method.

File: tests/render/test_data_documentation_site_builder.py (hunk 7 above)
Comment (eugmandel) [reply tests/render/test_data_documentation_site_builder.py:12]: +1

File: tests/render/test_data_documentation_site_builder.py (hunk 8 above)
Comment (eugmandel) [reply tests/render/test_data_documentation_site_builder.py:8]: ouch - will remove :)

**Pull request merged**